src/v/wasm/cache.cc

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "wasm/cache.h"
+
+#include "wasm/logger.h"
+
+#include <seastar/coroutine/as_future.hh>
+
+namespace wasm {
+
+namespace {
+
+/**
+ * Allows sharing an engine between multiple uses.
+ *
+ * Must live on a single core.
+ */
+class shared_engine : public engine {
+public:
+    explicit shared_engine(std::unique_ptr<engine> underlying)
+      : _underlying(std::move(underlying)) {}
+
+    ss::future<model::record_batch>
+    transform(model::record_batch batch, transform_probe* probe) override {
+        auto u = co_await _mu.get_units();
```

Comment:
Good question! It's possible to have as many waiters here as there are partitions on a core. So this doesn't scale with user input, but with the number of independent processors running. Long running transforms will be throttled and will yield to the scheduler, then eventually timeout (this is functionality will come in later PRs). So this is a risk but fundamentally a wasm engine can only execute a single thing at a time so this needs to be here. There is a fundamental tradeoff of being able to use more engines (and more memory) to allow for more concurrent transforms running, but we're still pinned to a single core. This means that we can't really get more throughput unless there is an async host call. The only true async host call is for schema registry (which is heavily cached inside the VM), all other yielding is only to prevent wasm from blocking other more important work.

From a product prospective, multiplexing these engines gives users a more explainable resource model (just assume you have 1 function instance per core for each function you deploy), and allows us to prevent wasm from consuming too many resources.

Reply:
>Long running transforms will be throttled and will yield to the scheduler, then eventually timeout

Cool, figured as much. Halting problem be damned!

>multiplexing these engines gives users a more explainable resource model

👍 👍